tox.ini

```diff
@@ -1,2 +1,197 @@
+deps = flake8
+
+[testenv:py{36,37,38,39,310,311}-test-{lin,win,mac}]
+description = Run unit tests with pytest
+labels =
+    test
+setenv =
+    mac: OPEN_CMD = {env:OPEN_CMD:open}
+    lin: OPEN_CMD = {env:OPEN_CMD:xdg-open}
+commands_pre =
+    - {tty:rm -f "{toxworkdir}{/}cov_report_{env_name}_html{/}index.html":true}
+    - {tty:rm -f "{toxworkdir}{/}pytest_report_{env_name}.html":true}
+commands =
+    pytest \
+    -p no:cacheprovider \
+    --verbose \
+    --doctest-modules \
+    --capture=fd \
+    --cov=patroni \
+    --cov-report=term-missing \
+    --cov-append \
+    {tty::--cov-report="xml\:{toxworkdir}{/}cov_report.{env_name}.xml"} \
+    {tty:--cov-report="html\:{toxworkdir}{/}cov_report_{env_name}_html":} \
+    {tty:--html="{toxworkdir}{/}pytest_report_{env_name}.html":} \
+    {posargs:tests patroni}
+commands_post =
+    - {tty:{env:OPEN_CMD} "{toxworkdir}{/}cov_report_{env_name}_html{/}index.html":true}
+    - {tty:{env:OPEN_CMD} "{toxworkdir}{/}pytest_report_{env_name}.html":true}
+deps =
+    -r requirements.txt
+    mock>=2.0.0
+    pytest
+    pytest-cov
+    pytest-html
+    py{37,38,39,310,311}-test-{lin,win}: psycopg[binary]
+    test-mac: psycopg2-binary
+    py36-test: psycopg2-binary
+platform =
+    lin: linux
+    mac: darwin
+    win: win32
+allowlist_externals =
+    rm
+    {env:OPEN_CMD}
+
+[testenv:dep]
+description = Check package dependency problems
+commands = pipdeptree -w fail
+deps =
+    -r requirements.txt
+    pipdeptree
+
+[testenv:type]
+description = Run static type checking
+deps = mypy
+commands = mypy {posargs:patroni}
+
+[testenv:black]
+description: Reformat code with black
+deps = black
+commands = black {posargs:patroni tests}
+
+[testenv:docker-build]
+description = Build docker containers needed for testing
+labels =
+    behave
+commands =
+    docker build . -t patroni -f Dockerfile
+    docker build . -t patroni-dev -f features/Dockerfile
+allowlist_externals = docker
+
+[testenv:docker-cmd]
+description = Run an arbitrary command in patroni-dev docker container
+commands_pre =
+    docker run \
+        -e DCS=etcd \
+        --name tox-{env_name}-{env:PYTHONHASHSEED} \
+        --detach \
+        --rm \
+        patroni-dev
+commands =
+    docker exec tox-{env_name}-{env:PYTHONHASHSEED} {posargs:echo "Hello world"}
+commands_post =
+    docker stop tox-{env_name}-{env:PYTHONHASHSEED}
+allowlist_externals = docker
+
+;[testenv:behave-etcd-{lin,win,mac}]
+[testenv:docker-behave-{etcd}]
+description = Run behaviour tests im patroni-dev docker container
+setenv =
+    etcd: DCS=etcd
+    CONTAINER_NAME = tox-{env_name}-{env:PYTHONHASHSEED}
+labels =
+    behave
+depends =
+    docker-build
+commands_pre =
+    docker run \
+        -v {tox_root}:/src \
+        -e DCS={env:DCS} \
+        --name {env:CONTAINER_NAME} \
+        --detach \
+        --rm \
+        {env:PATRONI_DEV_IMAGE:patroni-dev}
+    docker exec {env:CONTAINER_NAME} cp -a /src /home/postgres/src
```

Comment:
can we avoid copying `.tox/` here? otherwise at least I have some problems with "too many levels of symbolic links" for python interpreters inside it

Reply:
Yep, I'll add rsync to do the copy